Adding nrf52811 support

jakub-uC opened the pull request:
This PR adds support for nrf52811 SoC and DK.
The nRF52811 is a Bluetooth 5.1 Direction Finding SoC with comprehensive protocol support.
The nRF52811 SoC is capable of the latest features of Bluetooth 5.1, the most prominent being Direction Finding.

The radio in the nRF52811 SoC has comprehensive protocol capabilities, including Bluetooth 5.1 Directing Finding, all Bluetooth 5 features, 802.15.4, Thread, Zigbee, ANT and 2.4 GHz proprietary.
It has 4 dBm TX power and has been optimized to offer the best RX sensitivity of all SoCs in the nRF52 series.

The nRF52840 DK: PCA10056 is the recommend development kit, it emulates the nRF52811, and can be used as a starting point for development before moving over to a custom board. Please note that this development kit does not support Bluetooth Direction Finding. What is more it cannot be used with most of Arduino shields because of PCA10056 PIN layout.

Commit:
arm: nrf: Add support for the nRF52811

This commit adds basic support for nrf52811 in the arch SoC, dts
and nrfx folders.

The nRF52811 is a Bluetooth 5.1 Direction Finding SoC with comprehensive
protocol support.
The nRF52811 SoC is capable of the latest features of Bluetooth 5.1,
the most prominent being Direction Finding.

The radio in the nRF52811 SoC has comprehensive protocol capabilities,
including Bluetooth 5.1 Directing Finding, all Bluetooth 5 features,
802.15.4, Thread, Zigbee, ANT and 2.4 GHz proprietary.
It has 4 dBm TX power and has been optimized to offer the best RX
sensitivity of all SoCs in the nRF52 series.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
boards: arm: Add Nordic nRF52811 board

This commit adds support for board: nrf52811_PCA10056.

The nRF52840 DK: PCA10056 is the recommend development kit, it emulates
the nRF52811, and can be used as a starting point for development
before moving over to a custom board.

Please note that this development kit does not support Bluetooth
Direction Finding. What is more it cannot be used with most of Arduino
shields because of PCA10056 PIN layout.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>
Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

lemrey commented on boards/arm/nrf52811_pca10056/nrf52811_pca10056.dts:
```diff
@@ -0,0 +1,154 @@
+		};
+		scratch_partition: partition@da000 {
+			label = "image-scratch";
+			reg = <0x000da000 0x0001e000>;
```
Hi @jarz-nordic , `0xda000` is not correct, the chip only has `0x30000` bytes of flash. 
The previous partition ranges from `0x19000` to `0x26000` (`0xd000` bytes long), so we have `0x4000` bytes available, enough to fit a `storage_partition` but not a `scratch_partition` (for MCUboot).

jarz-nordic replied:
thank you for finding. I will update it.

jarz-nordic replied:
fixed

dbkinder commented on boards/arm/nrf52811_pca10056/doc/index.rst:
```diff
@@ -0,0 +1,25 @@
+The nrf52811_pca10056 board is a modified version of the
+:ref:`nrf52840_pca10056` that enforces the limitations imposed by the nRF52811
+IC, which is a variant of the original nRF52811 with radio and few peripherals
+from nRF52840 and. Since Nordic does not offer a development kit for the
```
and ... ?

jarz-nordic replied:
fixed

lemrey reviewed (commented):
Hi @jarz-nordic, the `storage_partition` is generally smaller than the `scratch_partition`. We should try and keep those proportions in this board, unless there are special requirements  (that I am unaware of :)

I looked at `nrf52840_pca10056` for reference, and suggest the following values which are scaled down by 1/6 (1024 -> 192) and rounded up to page boundaries, to maintain the same proportions:
- `scratch_partition` `0x5000` (20k)
- `storage_partition` `0x2000` (8k)

overheat commented:
@jarz-nordic Thank yo for adding nrf52811 support. I have pulled the code and try to build samples/bluetooth/beacon/ but failed.
Looks subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf52811.h is missing.

jarz-nordic commented:
@overheat : Thank you! I will update it.

ioannisg commented on boards/arm/nrf52811_pca10056/nrf52811_pca10056.dts:
```diff
@@ -0,0 +1,158 @@
+
+/ {
+	model = "Nordic PCA10056 Dev Kit";
+	compatible = "nordic,pca10056-dk", "nordic,nrf52811-qiaa",
```
Shouldn't this be -qfaa?

jarz-nordic replied:
fixed

ioannisg reviewed (changes requested):
Did a pass, looks very good.
I left a minor comment.

Did not review the timings in radio_nrf52811.h, I assume @cvinayak could do it.

jarz-nordic commented:
@ioannisg : Thank you for comments.
Regarding radio file I was adviced by @cvinayak to reuse timings from nrf52840.

ioannisg commented:
> 
> 
> @ioannisg : Thank you for comments.
> Regarding radio file I was adviced by @cvinayak to reuse timings from nrf52840.

Makes sense, it's similar Radio.

ioannisg commented:
@jarz-nordic you would need a license header in all files, to make the license check pass.

jarz-nordic commented:
@ioannisg : I added where it was possible. After discussion with @anangl we shall not add Apache license to `ext/hal/nordic`

ioannisg commented:
> 
> 
> @ioannisg : I added where it was possible. After discussion with @anangl we shall not add Apache license to `ext/hal/nordic`

You're right, we don't add that one in `ext/nordic` :)

Some headers are still missing (e.g. `nrf52811_pca10056_defconfig` both lic and copyright) or not complete (e.g. `CMakeLists.txt` & `board.cmake`, copyright is missing)

jarz-nordic commented:
@ioannisg : done

ioannisg reviewed (approved):
Looks ok to me.
@anangl pls take a look, as well.

ioannisg commented:
@jarz-nordic perhaps sthg wrong with the counter driver (see CI error)

anangl commented on ext/hal/nordic/nrfx_config_nrf52811.h:
```diff
@@ -0,0 +1,2448 @@
+/**
```
There are a couple of things that are missing or need to be corrected in this file, like NVMC and TEMP sections, CLOCK_CONFIG_LF_SRC settings, and UART enabling. I'd recommend to take `nrfx_config_nrf52810.h` from the latest master as a starting point and only add SPI[M|S]1 related entries to it achieve the desired consistency.

jarz-nordic replied:
done

cvinayak reviewed (approved):
Assuming that the controller changes are based on nRF52840 radio timings, the changes look good to me.

jarz-nordic commented on boards/arm/nrf52811_pca10056/Kconfig:
```diff
@@ -0,0 +1,19 @@
+# BT_CTLR depends on BT. When BT is enabled we should default to also
+# enabling the controller.
+config BT_CTLR
+	default y if BT
```
done

jarz-nordic commented on boards/arm/nrf52811_pca10056/nrf52811_pca10056.dts:
```diff
@@ -0,0 +1,158 @@
+};
+
+&pwm0 {
+	status = "ok";
```
done

jarz-nordic commented:
@anangl : Please take a look.

jarz-nordic commented:
@ioannisg I fixed CI test. It seems that this problem is also valid for some more Nordic DKs.

ioannisg commented:
Adding a "Block: HW Test" label, i.e. we should probably try to do some basic testing on HW before merging this Board/SOC support:
- sanity check on the board, kernel tests, etc.
- testing basic samples (e.g. Bluetooth, ...)

jarz-nordic commented:
@ioannisg : I've tested few samples on this hardware:

- /zephyr/samples/bluetooth/beacon/
- /zephyr/samples/bluetooth/peripheral_hr
- /zephyr/samples/hello_world
- /zephyr/samples/subsys/shell/shell_module

All tests were passed

jarz-nordic commented:
@ioannisg : counter test is failing on a board: holyiot_yj16019 which is not related to my PR.
The same tests is failing for nrf52810 which is already in Zephyr upstream. Apperantly there is some mess but in my opinion it shall not be fixed within this PR.

ioannisg commented:
> 
> 
> @ioannisg : counter test is failing for board: **holyiot_yj16019** which is not related to my PR.
> The same tests is failing for nrf52810 which is already in Zephyr upstream. Apperantly there is some mess but in my opinion it shall not be fixed within this PR.

Doesn't it compile, or compiles, runs bu fails?

ioannisg commented:
> 
> 
> @ioannisg : counter test is failing for board: **holyiot_yj16019** which is not related to my PR.
> The same tests is failing for nrf52810 which is already in Zephyr upstream. Apperantly there is some mess but in my opinion it shall not be fixed within this PR.

Great. Suppose, you are using nrf52840 DKs as hardware?

jarz-nordic commented:
@ioannisg : Yes I am using nrf52840 DK to emulate nrf52811 (nrf52811_pca10056)
These tests were passed on board: -DBOARD=nrf52811_pca10056

- /zephyr/samples/bluetooth/beacon/
- /zephyr/samples/bluetooth/peripheral_hr
- /zephyr/samples/hello_world
- /zephyr/samples/subsys/shell/shell_module

It also passed test:

- tests/drivers/counter/counter_basic_api


Currently in CI test tests/drivers/counter/counter_basic_api is failing for DK **holyiot_yj16019** which is not related to my PR. My PR is not causing this error.

ioannisg commented:
> 
> 
> @ioannisg : Yes I am using nrf52840 DK to emulate nrf52811 (nrf52811_pca10056)
> These tests were passed on board: -DBOARD=nrf52811_pca10056
> 
>     * /zephyr/samples/bluetooth/beacon/
> 
>     * /zephyr/samples/bluetooth/peripheral_hr
> 
>     * /zephyr/samples/hello_world
> 
>     * /zephyr/samples/subsys/shell/shell_module
> 
> 
> It also passed test:
> 
>     * tests/drivers/counter/counter_basic_api
> 
> 
> Currently in CI test tests/drivers/counter/counter_basic_api is failing for DK **holyiot_yj16019** which is not related to my PR. My PR is not causing this error.

Might be an easy fix, though, let me investigate a bit, since, you say, it affects 52810, too.

nordic-krch commented:
regarding failing CI: you can remove `counter` capability from holyiot_yj16019 board (https://github.com/zephyrproject-rtos/zephyr/blob/b1f3825ea3fb07f89eb4daebb547dd2b6356a89b/boards/arm/holyiot_yj16019/holyiot_yj16019.yaml#L14). It's probably copypasted.

You could also add nrf52811_pca10056 configuration for counter test: https://github.com/zephyrproject-rtos/zephyr/tree/master/tests/drivers/counter/counter_basic_api/boards

jarz-nordic commented:
@nordic-krch : I've added nrf52811 to supported boards, but as mentioned before nrf52810 is also failing this test despite the fact nrf52810 board is available for a longer period of time.

Why this test is actually failing for other boards while my PR shall not affect them?

ioannisg commented:
> 
> 
> @nordic-krch : I've added nrf52811 to supported boards, but as mentioned before nrf52810 is also failing this test despite the fact nrf52810 board is available for a longer period of time.
> 
> Why this test is actually failing for other boards while my PR shall not affect them?

Please, add the nrf52810_pca10040 configuration as well.

I think failing is maybe due to how CI selects the (test/sample, board) combinations to build for each run.

carlescufi committed:
boards: holyiot_yj16019: Remove counter from supported

The supported field is used by sanitycheck to select which tests will be
run on a particular board. The issue with the counter drivers, and in
particular on nRF ICs, is that the timer peripherals are disabled by
default and need to be enabled on a per-board basis inside a .conf file.
Since this board doesn't have a .conf file in
tests/drivers/counter/counter_basic_api/boards/ the test cannot be run
on this board.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
> > @nordic-krch : I've added nrf52811 to supported boards, but as mentioned before nrf52810 is also failing this test despite the fact nrf52810 board is available for a longer period of time.
> > Why this test is actually failing for other boards while my PR shall not affect them?
> 
> Please, add the nrf52810_pca10040 configuration as well.
> 
> I think failing is maybe due to how CI selects the (test/sample, board) combinations to build for each run.

Let's do that in a follow-up PR.

carlescufi commented on tests/drivers/counter/counter_basic_api/boards/nrf52811_pca10056.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_COUNTER_TIMER1_PRESCALER=4
+CONFIG_COUNTER_TIMER2=y
+CONFIG_COUNTER_TIMER2_PRESCALER=4
+CONFIG_COUNTER_RTC0=y
```
@nordic-krch @jarz-nordic @ioannisg @anangl I removed `CONFIG_COUNTER_RTC2=y` from here since there's only 2 RTCs in the 8211

jarz-nordic replied:
@carlescufi : Did you remove RTC2 or RTC1 ? Maybe I accidently set RTC2 instead of RTC1. Currently I do not see here RTC1 despite the fact nrf52811 has it.

nordic-krch replied:
RTC1 is used by the kernel.

ioannisg commented:
@carlescufi @galak , I suppose this could be merged now? The license check fail is expected.

nordic-krch commented:
+1 for merging it as it will fix CI for other PRs related to counter.
